File: src/test/java/org/saigon/striker/service/UserServiceTest.java
package org.saigon.striker.service;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.saigon.striker.model.UserEntity;
import org.saigon.striker.model.UserRepository;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.crypto.password.PasswordEncoder;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
import static org.assertj.core.api.Assertions.assertThatNullPointerException;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class UserServiceTest {

    private static final String ID = "userId";
    private static final String USERNAME = "username";
    private static final String PASSWORD = "password";
    private static final UserEntity USER_ENTITY = new UserEntity(ID, USERNAME, PASSWORD);

    @Captor
    private ArgumentCaptor<UserEntity> userEntityCaptor;
    @Mock
    private UserRepository userRepository;
    @Mock
    private PasswordEncoder passwordEncoder;
    @InjectMocks
    private UserService userService;

    @Test
    public void findByUsername() {
        when(userRepository.findByUsername(USERNAME)).thenReturn(Mono.just(USER_ENTITY));

        StepVerifier.create(userService.findByUsername(USERNAME))
                .assertNext(userDetails -> {
                    assertThat(userDetails.getUsername()).isEqualTo(USERNAME);
                    assertThat(userDetails.getPassword()).isEqualTo(PASSWORD);
                    assertThat(userDetails.getAuthorities())
                            .extracting(GrantedAuthority::getAuthority).containsOnly("ROLE_USER");
                })
                .verifyComplete();
    }

    @Test
    public void findByUsernameNotFound() {
        when(userRepository.findByUsername(USERNAME)).thenReturn(Mono.empty());

        StepVerifier.create(userService.findByUsername(USERNAME))
                .verifyComplete();
    }

    @Test
    public void findByUsernameInvalid() {
        assertThatNullPointerException().isThrownBy(() -> userService.findByUsername(null));
        assertThatIllegalArgumentException().isThrownBy(() -> userService.findByUsername(""));
    }

    @Test
    public void createUser() {
        when(passwordEncoder.encode(PASSWORD)).thenReturn("encodedPassword");
        when(userRepository.save(any())).thenReturn(Mono.just(USER_ENTITY));

        StepVerifier.create(userService.createUser(USER_ENTITY))
                .assertNext(this::assertUserEntity)
                .verifyComplete();

        verify(userRepository).save(userEntityCaptor.capture());
        assertThat(userEntityCaptor.getValue().getUsername()).isEqualTo(USERNAME);
        assertThat(userEntityCaptor.getValue().getPassword()).isEqualTo("encodedPassword");
    }

    @Test
    public void createUserInvalid() {
        assertThatNullPointerException().isThrownBy(() -> userService.createUser(null));
    }

    @Test
    public void getUser() {
        when(userRepository.findById(ID)).thenReturn(Mono.just(USER_ENTITY));

        StepVerifier.create(userService.getUser(ID))
                .assertNext(this::assertUserEntity)
                .verifyComplete();
    }

    @Test
    public void getUserNotFound() {
        when(userRepository.findById(ID)).thenReturn(Mono.empty());

        StepVerifier.create(userService.getUser(ID))
                .verifyComplete();
    }

    @Test
    public void getUserInvalid() {
        assertThatNullPointerException().isThrownBy(() -> userService.getUser(null));
        assertThatIllegalArgumentException().isThrownBy(() -> userService.getUser(""));
    }

    @Test
    public void deleteUser() {
        when(userRepository.deleteById(ID)).thenReturn(Mono.empty());

        StepVerifier.create(userService.deleteUser(ID))
                .verifyComplete();
    }

    @Test
    public void deleteUserInvalid() {
        assertThatNullPointerException().isThrownBy(() -> userService.deleteUser(null));
        assertThatIllegalArgumentException().isThrownBy(() -> userService.deleteUser(""));
    }

    private void assertUserEntity(UserEntity userEntity) {
        assertThat(userEntity.getId()).isEqualTo(ID);
        assertThat(userEntity.getUsername()).isEqualTo(USERNAME);
        assertThat(userEntity.getPassword()).isEqualTo(PASSWORD);
    }
}

File: src/test/kotlin/org/saigon/striker/service/UserServiceTest.kt
package org.saigon.striker.service

import io.mockk.Runs
import io.mockk.coEvery
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.just
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.NullSource
import org.saigon.striker.UserEntitySource
import org.saigon.striker.model.UserEntity
import org.saigon.striker.model.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import reactor.core.publisher.Mono

@ExtendWith(MockKExtension::class)
class UserServiceTest {

    @MockK
    lateinit var userRepository: UserRepository

    @MockK
    lateinit var passwordEncoder: PasswordEncoder

    @InjectMockKs
    lateinit var userService: UserService

    @NullSource
    @UserEntitySource
    @ParameterizedTest
    fun `find user details by username`(user: UserEntity?) {
        every { userRepository.findByUsernameMono(user?.username ?: "unknown") } returns Mono.justOrEmpty(user)

        val userDetails = userService.findByUsername(user?.username ?: "unknown").block()

        if (user == null) {
            assertThat(userDetails).isNull()
        } else {
            assertThat(userDetails?.username).isEqualTo(user.username)
            assertThat(userDetails?.password).isEqualTo(user.password)
            assertThat(userDetails?.authorities).extracting("authority").containsOnly("ROLE_USER")
        }
    }

    @UserEntitySource
    @ParameterizedTest
    fun `create user`(user: UserEntity) = runBlocking<Unit> {
        coEvery { userRepository.findByUsername(user.username) } returns null
        coEvery { userRepository.save(user.copy(password = "encodedPassword")) } returns user
        every { passwordEncoder.encode(user.password) } returns "encodedPassword"

        val result = userService.createUser(user)

        assertThat(result).isEqualTo(user)
    }

    @UserEntitySource
    @ParameterizedTest
    fun `create existing user`(user: UserEntity) {
        coEvery { userRepository.findByUsername(user.username) } returns user

        val exception = assertThrows<UsernameAlreadyExistsException> {
            runBlocking { userService.createUser(user) }
        }

        assertThat(exception).hasMessageContaining("Username '${user.username}' already exists")
    }

    @NullSource
    @UserEntitySource
    @ParameterizedTest
    fun `get user by id`(user: UserEntity?) = runBlocking<Unit> {
        coEvery { userRepository.findById(user?.id ?: "unknown") } returns user

        val result = userService.getUser(user?.id ?: "unknown")

        assertThat(result).isEqualTo(user)
    }

    @NullSource
    @UserEntitySource
    @ParameterizedTest
    fun `get user by username`(user: UserEntity?) = runBlocking<Unit> {
        coEvery { userRepository.findByUsername(user?.username ?: "unknown") } returns user

        val result = userService.getUserByUsername(user?.username ?: "unknown")

        assertThat(result).isEqualTo(user)
    }

    @Test
    fun `delete user by id`() = runBlocking<Unit> {
        coEvery { userRepository.deleteById("id") } just Runs

        userService.deleteUser("id")
    }
}
